In [1]:
import graphlab as gl

A newer version of GraphLab Create (v1.8.1) is available! Your current version is v1.8.

You can use pip to upgrade the graphlab-create package. For more information see https://dato.com/products/create/upgrade.


In [2]:
song_data = gl.SFrame('song_data.gl/') 

[INFO] This non-commercial license of GraphLab Create is assigned to email.jrv@gmail.com and will expire on January 20, 2017. For commercial licensing options, visit https://dato.com/buy/.

[INFO] Start server at: ipc:///tmp/graphlab_server-3270 - Server binary: /home/jrv/anaconda2/envs/dato-env/lib/python2.7/site-packages/graphlab/unity_server - Server log: /tmp/graphlab_server_1454036612.log
[INFO] GraphLab Server Version: 1.8


In [9]:
gl.canvas.set_target('ipynb')

In [12]:
song_data.head()

Columns:
	user_id	str
	song_id	str
	listen_count	int
	title	str
	artist	str
	song	str

Rows: 10

Data:
+-------------------------------+--------------------+--------------+
|            user_id            |      song_id       | listen_count |
+-------------------------------+--------------------+--------------+
| b80344d063b5ccb3212f76538f... | SOAKIMP12A8C130995 |      1       |
| b80344d063b5ccb3212f76538f... | SOBBMDR12A8C13253B |      2       |
| b80344d063b5ccb3212f76538f... | SOBXHDL12A81C204C0 |      1       |
| b80344d063b5ccb3212f76538f... | SOBYHAJ12A6701BF1D |      1       |
| b80344d063b5ccb3212f76538f... | SODACBL12A8C13C273 |      1       |
| b80344d063b5ccb3212f76538f... | SODDNQT12A6D4F5F7E |      5       |
| b80344d063b5ccb3212f76538f... | SODXRTY12AB0180F3B |      1       |
| b80344d063b5ccb3212f76538f... | SOFGUAY12AB017B0A8 |      1       |
| b80344d063b5ccb3212f76538f... | SOFRQTD12A81C233C0 |      1       |
| b80344d063b5ccb3212f76538f... | SOHQWYZ12A6D4FA701 |      1       |
+-------------------------------+--------------------+--------------+
+-------------------------------+-------------------------------+
|             title             |             artist            |
+-------------------------------+-------------------------------+
|            The Cove           |          Jack Johnson         |
|        Entre Dos Aguas        |         Paco De Lucia         |
|            Stronger           |           Kanye West          |
|         Constellations        |          Jack Johnson         |
|          Learn To Fly         |          Foo Fighters         |
|  Apuesta Por El Rock 'N' Roll |      Héroes del Silencio      |
|         Paper Gangsta         |           Lady GaGa           |
|         Stacked Actors        |          Foo Fighters         |
|         Sehr kosmisch         |            Harmonia           |
| Heaven's gonna burn your eyes | Thievery Corporation feat.... |
+-------------------------------+-------------------------------+
+-------------------------------+
|              song             |
+-------------------------------+
|    The Cove - Jack Johnson    |
| Entre Dos Aguas - Paco De ... |
|     Stronger - Kanye West     |
| Constellations - Jack Johnson |
|  Learn To Fly - Foo Fighters  |
| Apuesta Por El Rock 'N' Ro... |
|   Paper Gangsta - Lady GaGa   |
| Stacked Actors - Foo Fighters |
|    Sehr kosmisch - Harmonia   |
| Heaven's gonna burn your e... |
+-------------------------------+
[10 rows x 6 columns]

In [19]:
Kanye_users = song_data[song_data['artist'] == 'Kanye West']

In [20]:
Kanye_users.head()

Columns:
	user_id	str
	song_id	str
	listen_count	int
	title	str
	artist	str
	song	str

Rows: 10

Data:
+-------------------------------+--------------------+--------------+
|            user_id            |      song_id       | listen_count |
+-------------------------------+--------------------+--------------+
| b80344d063b5ccb3212f76538f... | SOBXHDL12A81C204C0 |      1       |
| b80344d063b5ccb3212f76538f... | SOMLMKI12A81C204BC |      1       |
| 5d5e0142e54c3bb7b69f548c2e... | SORFASW12A81C22AE7 |      2       |
| 537340ff896dea11328910013c... | SOBXHDL12A81C204C0 |      2       |
| 7dd192c8bd4f27f573cb15e865... | SOOLPFK12A58A7BDE3 |      5       |
| 8fce200f3912e9608e3b1463cd... | SOBXHDL12A81C204C0 |      2       |
| 8fce200f3912e9608e3b1463cd... | SOIBSWV12A6D4F6AB3 |      1       |
| a56bf59af6edc5ae6c92d61ddd... | SONGNHO12AB0183915 |      1       |
| 8fa25e588aeedaa539674babb7... | SOOLPFK12A58A7BDE3 |      1       |
| e8612acfb1572297ea0eaaa1f2... | SOIYWPZ12A81C204EF |      2       |
+-------------------------------+--------------------+--------------+
+------------------+------------+-------------------------------+
|      title       |   artist   |              song             |
+------------------+------------+-------------------------------+
|     Stronger     | Kanye West |     Stronger - Kanye West     |
|     Champion     | Kanye West |     Champion - Kanye West     |
|     Stronger     | Kanye West |     Stronger - Kanye West     |
|     Stronger     | Kanye West |     Stronger - Kanye West     |
| Flashing Lights  | Kanye West |  Flashing Lights - Kanye West |
|     Stronger     | Kanye West |     Stronger - Kanye West     |
| Through The Wire | Kanye West | Through The Wire - Kanye West |
|     Bad News     | Kanye West |     Bad News - Kanye West     |
| Flashing Lights  | Kanye West |  Flashing Lights - Kanye West |
|    Homecoming    | Kanye West |    Homecoming - Kanye West    |
+------------------+------------+-------------------------------+
[10 rows x 6 columns]

In [21]:
users = Kanye_users['user_id'].unique()

In [22]:
len(users)

2522

## Kanye west users : 2522

In [25]:
artist_freq = song_data.groupby(key_columns='artist', operations={'total_count': gl.aggregate.SUM('listen_count')})

In [26]:
artist_freq = artist_freq.sort('total_count', ascending = False )

In [27]:
artist_freq.head()

Columns:
	artist	str
	total_count	int

Rows: 10

Data:
+------------------------+-------------+
|         artist         | total_count |
+------------------------+-------------+
|     Kings Of Leon      |    43218    |
|     Dwight Yoakam      |    40619    |
|         Björk          |    38889    |
|        Coldplay        |    35362    |
| Florence + The Machine |    33387    |
|     Justin Bieber      |    29715    |
|    Alliance Ethnik     |    26689    |
|      OneRepublic       |    25754    |
|         Train          |    25402    |
|     The Black Keys     |    22184    |
+------------------------+-------------+
[10 rows x 2 columns]

In [28]:
artist_freq[-1]

{'artist': 'William Tabbert', 'total_count': 14}

In [43]:
artist_freq[artist_freq['artist'] =='Kanye West']

Columns:
	artist	str
	total_count	int

Rows: Unknown

Data:
+------------+-------------+
|   artist   | total_count |
+------------+-------------+
| Kanye West |     9992    |
+------------+-------------+
[? rows x 2 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use len(sf) to force materialization.

In [44]:
artist_freq[artist_freq['artist'] =='Foo Fighters']

Columns:
	artist	str
	total_count	int

Rows: Unknown

Data:
+--------------+-------------+
|    artist    | total_count |
+--------------+-------------+
| Foo Fighters |     9504    |
+--------------+-------------+
[? rows x 2 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use len(sf) to force materialization.

In [45]:
artist_freq[artist_freq['artist'] =='Taylor Swift']

Columns:
	artist	str
	total_count	int

Rows: Unknown

Data:
+--------------+-------------+
|    artist    | total_count |
+--------------+-------------+
| Taylor Swift |    19376    |
+--------------+-------------+
[? rows x 2 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use len(sf) to force materialization.

In [46]:
artist_freq[artist_freq['artist'] =='Lady GaGa']

Columns:
	artist	str
	total_count	int

Rows: Unknown

Data:
+-----------+-------------+
|   artist  | total_count |
+-----------+-------------+
| Lady GaGa |    12224    |
+-----------+-------------+
[? rows x 2 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use len(sf) to force materialization.

### Most Frequent :  Kings Of Leon 	: 43218
### Least Frequent : William Tabbert : 14

In [30]:
train_data,test_data = song_data.random_split(.8,seed=0)

In [32]:
personalized_model = gl.item_similarity_recommender.create(train_data,
                                                                user_id='user_id',
                                                                item_id='song')

PROGRESS: Recsys training: model = item_similarity
PROGRESS: Warning: Ignoring columns song_id, listen_count, title, artist;
PROGRESS:     To use one of these as a target column, set target = <column_name>
PROGRESS:     and use a method that allows the use of a target.
PROGRESS: Preparing data set.
PROGRESS:     Data has 893580 observations with 66085 users and 9952 items.
PROGRESS:     Data prepared in: 0.969737s
PROGRESS: Computing item similarity statistics:
PROGRESS: Computing most similar items for 9952 items:
PROGRESS: +-----------------+-----------------+
PROGRESS: | Number of items | Elapsed Time    |
PROGRESS: +-----------------+-----------------+
PROGRESS: | 1000            | 1.52999         |
PROGRESS: | 2000            | 1.58671         |
PROGRESS: | 3000            | 1.65018         |
PROGRESS: | 4000            | 1.70529         |
PROGRESS: | 5000            | 1.75694         |
PROGRESS: | 6000            | 1.80686         |
PROGRESS: | 7000            | 1.85607         |

In [33]:
subset_test_users = test_data['user_id'].unique()[0:10000]

In [36]:
test_songs = personalized_model.recommend(subset_test_users,k=1)

PROGRESS: recommendations finished on 1000/10000 queries. users per second: 1966.17
PROGRESS: recommendations finished on 2000/10000 queries. users per second: 2096.39
PROGRESS: recommendations finished on 3000/10000 queries. users per second: 2129.85
PROGRESS: recommendations finished on 4000/10000 queries. users per second: 2150.9
PROGRESS: recommendations finished on 5000/10000 queries. users per second: 2157.93
PROGRESS: recommendations finished on 6000/10000 queries. users per second: 2158.06
PROGRESS: recommendations finished on 7000/10000 queries. users per second: 2156.4
PROGRESS: recommendations finished on 8000/10000 queries. users per second: 2155.7
PROGRESS: recommendations finished on 9000/10000 queries. users per second: 2145.74
PROGRESS: recommendations finished on 10000/10000 queries. users per second: 2143.25


In [38]:
subset_test_user_songs = test_songs.groupby(key_columns='song', operations={'count': gl.aggregate.COUNT()})

In [39]:
subset_test_user_songs.head()

Columns:
	song	str
	count	int

Rows: 10

Data:
+-------------------------------+-------+
|              song             | count |
+-------------------------------+-------+
|    The Climb - Miley Cyrus    |   3   |
| Hey Daddy (Daddy's Home) -... |   6   |
| I Gotta Feeling - Black Ey... |   13  |
|        Cerdo - Molotov        |   1   |
|      Ich Will - Rammstein     |   10  |
|      Too Deep - Girl Talk     |   3   |
|      Dumpweed - Blink-182     |   2   |
|   Guys Like Me - Eric Church  |   2   |
| I Can't Love You Back - Ea... |   2   |
|         Freedom - Akon        |   4   |
+-------------------------------+-------+
[10 rows x 2 columns]

In [41]:
subset_test_user_songs = subset_test_user_songs.sort('count', ascending = False )

In [42]:
subset_test_user_songs.head()

Columns:
	song	str
	count	int

Rows: 10

Data:
+--------------------------------+-------+
|              song              | count |
+--------------------------------+-------+
|          Undo - Björk          |  431  |
|     Secrets - OneRepublic      |  386  |
|    Revelry - Kings Of Leon     |  232  |
| You're The One - Dwight Yoakam |  170  |
| Fireflies - Charttraxx Karaoke |  122  |
|    Hey_ Soul Sister - Train    |  107  |
| Horn Concerto No. 4 in E f...  |   97  |
|    Sehr kosmisch - Harmonia    |   71  |
| OMG - Usher featuring will...  |   59  |
| Dog Days Are Over (Radio E...  |   53  |
+--------------------------------+-------+
[10 rows x 2 columns]


### Most Reccomended Song : Undo - Björk :	431